[RNMobile] Fix undo/redo history when inserting a link configured to open in a new tab

derekblank opened the pull request:
## What?
Fixes a bug affecting undo/redo functionality on mobile. When a link is configured to open in a new tab and text is typed after this link, it is not possible to undo and then use the Redo button to redo the text that was added after the link.

**Original issue:**
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/2091

**Related issues:**
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/1160
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/1161

**Gutenberg Mobile PR:**
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5747

## Why?
There are several related issues that describe the broken behavior with undo/redo when using the "Open in new tab" option on mobile. One symptom of this issue is that an extra state event occurs on mobile when a link is set to open in a new tab. This is demonstrated visually in https://github.com/wordpress-mobile/gutenberg-mobile/issues/2091#issuecomment-662212872 with further information. Configuring a link to open in a new tab shifts the `rel` attribute within the HTML, which causes the [edit reducer](https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/reducer.js#L523-L558) to disrupt the redo history.

Example HTML when a link is created without opening in a new tab:

```
<a href="https://wordpress.com" target="_blank" rel="noreferrer noopener">WordPress.com</a>
```

Example HTML when a link is created and configured to open in a new tab:

```
<a rel="noreferrer noopener" href="https://wordpress.com" target="_blank">WordPress.com</a>
```

## How?
This PR adds `rel` to the object of attributes passed to the link formatting, along with `url` and `target`:

```
attributes: {
	url: 'href',
	target: 'target',
	rel: 'rel',
},
```

When `rel` is passed with the attributes object to [registerFormatType](https://github.com/WordPress/gutenberg/blob/trunk/packages/format-library/src/index.js#L12), the `rel` attribute is no longer an unassigned attribute, and does not shift places when the HTML attributes are re-serialized. Therefore, the extra state event is not fired on mobile. On the web, the extra state event is not fired in either circumstance, and the re-serialization of the attributes has no effect on behavior. 

## Testing Instructions

1. In a post, add a paragraph block and type some text.
2. Create a link using the Link button in the Bottom Bar.
3. When configuring the link options, toggle the switch to **Open in new tab**.
4. Switch to HTML mode, and note the position of the `rel` attribute within the link's HTML. (It should be the final attribute.)
5. Switch back to Visual mode and continue typing a few more words of text.
6. Test the Undo and Redo buttons in the Bottom Bar.
7. Use the Undo button and step backward to at least the point before the link was created.
8. Use the Redo button to step forward to the point where the link was created, and also step forward to the text that was typed after the link.

**Demo builds for testing:**
- iOS: https://github.com/wordpress-mobile/WordPress-iOS/pull/20661
- Android: https://github.com/wordpress-mobile/WordPress-Android/pull/18405

## Screenshots

https://github.com/WordPress/gutenberg/assets/643285/bae80218-275c-4970-9e05-a37bb57088b6

derekblank committed:
Add rel to link formatting attributes

derekblank committed:
Update CHANGELOG to add undo/redo fix to Unreleased section

derekblank committed:
Merge remote-tracking branch 'origin' into rnmobile/fix/undo-redo-history

derekblank committed:
Merge trunk

derekblank committed:
Merge remote-tracking branch 'origin' into rnmobile/fix/undo-redo-history

twstokes commented on packages/react-native-editor/CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ For each user feature we should also add a importance categorization label  to i
 -   [*] Fix crash when trying to convert to regular blocks an undefined/deleted reusable block [#50475]
 -   [**] Tapping on a nested block now gets focus directly instead of having to tap multiple times depeding on the nesting levels. [#50108]
 -   [*] Use host app namespace in reusable block message [#50478]
+-   [**] Fix bug affecting undo and redo history when instering a link configured to open in a new tab [#50460]
```
```suggestion
-   [**] Fix bug affecting undo and redo history when inserting a link configured to open in a new tab [#50460]
```

derekblank committed:
Update packages/react-native-editor/CHANGELOG.md

Co-authored-by: Tanner Stokes <tanner.stokes@automattic.com>

dcalhoun commented on packages/react-native-editor/CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ For each user feature we should also add a importance categorization label  to i
 -   [*] Fix crash when trying to convert to regular blocks an undefined/deleted reusable block [#50475]
 -   [**] Tapping on a nested block now gets focus directly instead of having to tap multiple times depeding on the nesting levels. [#50108]
 -   [*] Use host app namespace in reusable block message [#50478]
+-   [**] Fix bug affecting undo and redo history when inserting a link configured to open in a new tab [#50460]
```
Given these are notes intended to be user-facing, do you think it would be beneficial to add more details regarding the fixed bug? 

```suggestion
-   [**] Configuring a link to open in a new tab no longer results in a partial loss of edit history (undo and redo) [#50460]
```

dcalhoun reviewed (approved):
Super interesting! Thank you for researching and addressing this. 

The prototype iOS build does not pass the test plan. However, my guess is this occurs because the prototype build does not include the fixed code. The https://github.com/wordpress-mobile/gutenberg-mobile/pull/5747 PR does not include a new bundle via `npm run bundle`. The Android prototype build succeeds, though. 👍🏻 

Would you be willing to follow up with tests covering this bug? In general, I think it would be good to always include a test with each bug we fix. For this specific bug, adding a test to the [Editor History file](https://github.com/WordPress/gutenberg/blob/99c3dda88168b262a65ceb33cd6dbce6c5e02b40/test/native/integration/editor-history.native.js) makes sense to me.

derekblank committed:
Update packages/react-native-editor/CHANGELOG.md

Co-authored-by: David Calhoun <github@davidcalhoun.me>

derekblank commented:
> The [wordpress-mobile/gutenberg-mobile#5747](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5747) PR does not include a new bundle via `npm run bundle`.

Thanks for pointing this out. There were a handful of unrelated CI issues on the Gutenberg Mobile PR, and I ended up manually updating the Gutenberg ref (possibly incorrectly), which may have caused it the branches to be out of sync. I'll update the iOS build. 👍 

> Would you be willing to follow up with tests covering this bug? In general, I think it would be good to always include a test with each bug we fix. For this specific bug, adding a test to the [Editor History file](https://github.com/WordPress/gutenberg/blob/99c3dda88168b262a65ceb33cd6dbce6c5e02b40/test/native/integration/editor-history.native.js) makes sense to me.

This is a good idea, and I'll start working on it to include before we merge this PR.

dcalhoun commented:
> > The [wordpress-mobile/gutenberg-mobile#5747](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5747) PR does not include a new bundle via `npm run bundle`.
> 
> Thanks for pointing this out. There were a handful of unrelated CI issues on the Gutenberg Mobile PR, and I ended up manually updating the Gutenberg ref (possibly incorrectly), which may have caused it the branches to be out of sync. I'll update the iOS build. 👍

You may already understand this, but your response gave me pause. The updated Gutenberg ref (git submodule commit SHA) is updated correctly. The missing piece is committing an updated app bundle produced via `npm run bundle` within `gutenberg-mobile`. A Buildkite CI tasks [publishes an artifact](https://github.com/wordpress-mobile/gutenberg-mobile/blob/bf65922627fbc390151344a262afbf33bdb4f6a4/.buildkite/publish-react-native-bridge-android-artifacts.sh) for Android, but there is no such CI task for iOS. In order to integrate an updated Gutenberg ref into a `WordPress-iOS` prototype build, we must run `npm run bundle` and commit the modified files.

derekblank commented:
> The missing piece is committing an updated app bundle produced via `npm run bundle` within `gutenberg-mobile`. A Buildkite CI tasks [publishes an artifact](https://github.com/wordpress-mobile/gutenberg-mobile/blob/bf65922627fbc390151344a262afbf33bdb4f6a4/.buildkite/publish-react-native-bridge-android-artifacts.sh) for Android, but there is no such CI task for iOS. In order to integrate an updated Gutenberg ref into a `WordPress-iOS` prototype build, we must run `npm run bundle` and commit the modified files.

No, I was actually not aware of that, but that makes much more sense now. I was aware of the `npm run bundle` command, but have only used it in the context of managing a release. Specifically, I didn't put together that there isn't a matching CI task for iOS. This is the first time I've needed/wanted to create a prototype build outside of creating a release, so thanks again for clarifying that. 

I was looking for documentation on this, and noted that it is mentioned in the [WPiOS Integration](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/WPiOS%20Integration.md#how-to-test-a-gutenberg-mobile-pr-in-wpios) doc. I had only read about it in the release context over in the [release-toolkit-gutenberg-mobile](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/blob/trunk/Releasing.md#making-a-release) docs. I test PRs in the host apps using the local gutenberg-mobile checkout, so it's possible that I once read the latter half of the WPiOS Integration doc and put it out of my mind. Do you think it would be beneficial to link or mention the host app Integration documentation from the Gutenberg Mobile README?

dcalhoun commented:
> Do you think it would be beneficial to link or mention the host app Integration documentation from the Gutenberg Mobile README?

Yeah, sounds like a good addition to the existing "Run" section. I think it makes sense for the README to default to the Demo editor, but a note could be added linking to documentation on how to integrate into the host WordPress iOS and Android apps.

derekblank committed:
Add test for checking undo/redo history when a link is configured to open in a new tab

derekblank committed:
Update link settings modal test case to check for correct editor history

derekblank committed:
Merge branch 'rnmobile/fix/undo-redo-history' of https://github.com/WordPress/gutenberg into rnmobile/fix/undo-redo-history

dcalhoun committed:
test: Fix React prop typo

dcalhoun committed:
test: Simplify test queries

The targeted elements do not require asynchronous queries. Using
synchronous queries is more concise and easier to comprehend.

dcalhoun committed:
test: Increase editor history link test accuracy

The targeted bug occurs when:

1. Changing a link to "Open in a new tab."
2. Performing additional changes, e.g. typing.
3. Undoing both actions.
4. Redoing both actions.

Thus, this changes the initial HTML, adds additional undo/redo
invocations, and updates inline snapshots to match the expected
outcomes.

derekblank committed:
Merge remote-tracking branch 'origin' into rnmobile/fix/undo-redo-history

derekblank commented:
@dcalhoun I've added a WIP for a [Editor History E2E test](https://github.com/WordPress/gutenberg/pull/50460/commits/c4bb30003698367981d4bcc837f9a796f9d7f523). I'm curious of your thoughts on this approach before proceeding further. Although just a scaffold at the moment, the test structure is similar to the [integration test](https://github.com/WordPress/gutenberg/pull/50460/files#diff-a663c09aa2aaa193aff09f58d69db6efca7fc1474e015221ae12b6a95e126124) in its arrangements and actions but with more realistic editor behavior, and could potentially replace the integration test in purpose. 

Specifically:
- **Test location**. I placed the Editor History E2E along with other E2E tests in the react-native-editor package, as opposed to placing it within `__device-tests__` in the root of Gutenberg Mobile. The Gutenberg Mobile E2E tests felt more snapshot-focused, while the react-native-editor E2E tests felt more assertion focused. I'm trying to understand if there is a more nuanced difference than that between the two test locations. The current test strategy I have outlined for Editor History E2E feels more assertion focused, but I am open to another perspective.
- **Test separation of concerns**. I've created a separate file for the Editor History E2E test. Would it make more sense to include as a test case in an existing E2E file, like gutenberg-editor-paragraph.test.js? Since it is intended to test editor history, I think it makes sense as its own file.
- **Selectors and helpers**. In the new [Editor History E2E](https://github.com/WordPress/gutenberg/blob/c4bb30003698367981d4bcc837f9a796f9d7f523/packages/react-native-editor/__device-tests__/gutenberg-editor-history.test.js) test file, I've left scaffolding comments where new E2E helpers and selectors would need to be implemented. If you have a moment to review, I'd be interested in any tips or caveats on these scaffolded areas -- particularly if you are able to point to any helpers that may already exist that would accomplish the described action (e.g., selecting the text input from a paragraph block, populating it with text, selecting a specific word of that text to "highlight" in order to press the Link button and invoke the Link Settings bottom sheet, etc. I believe most of these helpers/selectors will need to be created.)

dcalhoun commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -673,6 +673,25 @@ class EditorPage {
 		return isPresent;
 	}
 
+	async getParagraphBlockTextInputAtPosition( blockName, position = 1 ) {
```
Can we not reuse the existing [`getTextBlockAtPosition`](https://github.com/WordPress/gutenberg/blob/a2d829595a8af90b4fc29283c2d59bbfa11af64d/packages/react-native-editor/__device-tests__/pages/editor-page.js#L85-L98) helper? It is not clear to me the purpose or difference of this new helper.

derekblank replied:
This piece of code was intended to access the text input of the block in order to select a word, and there could be a more straightforward approach. It came from copying the way it was handled on the Shortcode block [here](https://github.com/WordPress/gutenberg/blob/0b70248e033d842468a450fda7f30efb8b8d121f/packages/react-native-editor/__device-tests__/pages/editor-page.js#L896-L913). Using `getTextBlockAtPosition` appeared to initiate the block for drag-and-drop, not focus the input. I was still assessing if this code was actually needed [here](https://github.com/WordPress/gutenberg/blob/c4bb30003698367981d4bcc837f9a796f9d7f523/packages/react-native-editor/__device-tests__/gutenberg-editor-history.test.js#L21). It seems like a more generic helper could be written to handle text inputs in the future.

dcalhoun commented on packages/react-native-editor/__device-tests__/gutenberg-editor-history.test.js:
```diff
@@ -0,0 +1,60 @@
+
+	// Select text
+	// Note: may need to write getParagraphBlockTextInputAtPosition helper to get the TextInput element
+	// Todo: how to select one single word of text from the input?
+	await selectTextFromElement( editorPage.driver, paragraphBlockElement );
```
If the hypothetical word is already a link, an alternative to selecting a single word would be merely placing the caret in the middle of said word. That will result in the formatting toolbar — i.e. the Link button — to be in a state to control the word. Moving the cursor could be a done by calculating, or "hard coding," the coordinates for the target word. 

We would need to either use an initial HTML containing a link or go through the steps to add a link. The latter would involve typing a bit, tapping the Link button, filling out link information, dismissing the bottom sheet, and typing some more. This would result in a sentence with a word(s) set as a link.

derekblank replied:
Good idea. I had been attempting to use [`clickMiddleOfElement`](https://github.com/WordPress/gutenberg/blob/0b70248e033d842468a450fda7f30efb8b8d121f/packages/react-native-editor/__device-tests__/helpers/utils.js#L287) to select a word. After reading this comment, I believe the drag-and-drop behavior I mentioned being initiated above may have been due to using `clickMiddleOfElement` when the broader paragraph block itself was selected (and not the text input within it).

dcalhoun commented on packages/react-native-editor/__device-tests__/gutenberg-editor-history.test.js:
```diff
@@ -0,0 +1,60 @@
+	// Test redo button
+	// Test redo button
+
+	// Verify the link attributes have not shifted with an assertion
```
To be sure I understand correctly, the focus of the test is to assert the entire history is re-playable, not attributes shifting, correct? If so, asserting the attribute position may not be necessary or already covered by integration/unit tests. That said, it may not be that "expensive" to assert this as well.

derekblank replied:
You are correct -- the assertion of the replayable history should be the focus of the test. The shifting attributes are a symptom of the issue that we could also check for with an assertion, but this may not be the best place for it.

dcalhoun reviewed (commented):
> The Gutenberg Mobile E2E tests felt more snapshot-focused, while the react-native-editor E2E tests felt more assertion focused. I'm trying to understand if there is a more nuanced difference than that between the two test locations.

The tests in the former location are either Automattic-centric (e.g. Jetpack) or experimental (e.g. visual snapshots). Additionally, you may note that not every test in latter location are run within the `gutenberg` CI; this was done due to how slow and flaky the suite is. Ideally, we would run all relevant tests within the `gutenberg` CI. Efforts to replace e2e tests with integration tests relate to this subject. 

Also, not every e2e test is _automatically_ run for every PR within `gutenberg-mobile`. You will note the "canary" name in certain e2e test files. These are run within the `gutenberg-mobile` CI for every PR. All other e2e tests are run when a person manually initiates the "Optional UI Tests" CI task in CircleCI. 

We are careful to add e2e tests that run by default in `gutenberg-mobile` and even more so within `gutenberg`. Placing this in `gutenberg` makes sense to me, but we should likely not run it in the CI. 

> Would it make more sense to include as a test case in an existing E2E file, like gutenberg-editor-paragraph.test.js? Since it is intended to test editor history, I think it makes sense as its own file.

I think this is largely a judgement call. It makes sense from a subject matter perspective that a single file focuses upon the feature of editor history. However, there is a test environment setup cost we pay for each test file. I.e. because e2e tests are so slow, it is beneficial to group several together rather than creating a new environment for each individual test. 

> If you have a moment to review, I'd be interested in any tips or caveats on these scaffolded areas -- particularly if you are able to point to any helpers that may already exist that would accomplish the described action

I left inline comments with any suggestions I have. However, keep in mind there are a lot of helpers and I am not intimately familiar with them all. I may have missed opportunities. My technique is to ["fold"](https://code.visualstudio.com/docs/editor/codebasics#_folding) all code in the `editor-page` and `utils` files to get a high-level view of the available helpers. 

<details><summary>Folded test helper code</summary>

![folded-test-helper-code](https://github.com/WordPress/gutenberg/assets/438664/796fbe1d-ff2d-4113-b4f9-b9f4c5b7c254)

</details>

Additionally, my recommendation would be to postpone creating new test helpers until we identify common patterns across several tests or a need to simplify a very complex or brittle interaction. As discussed, there are a lot of helpers; frequently adding more will continue to increase the incomprehensibility of them, reducing the usefulness of the helpers. 

Lastly — and I should have shared this in my original review — if we believe completing new test coverage will take a good amount of time, it may be worth merging this PR without tests and opening a separate PR adding tests. That might allow us to deliver a bug fix to users sooner.

derekblank commented:
@dcalhoun Thank you for reviewing, and for your explanations regarding the differences between the Gutenberg and Gutenberg Mobile CI tests, as well as the explanations for CI considerations. That all makes sense. 👍 

> Lastly — and I should have shared this in my original review — if we believe completing new test coverage will take a good amount of time, it may be worth merging this PR without tests and opening a separate PR adding tests. That might allow us to deliver a bug fix to users sooner.

I agree here too. After also discussing with @twstokes, I think we should pause on the E2E test due to complexities of writing additional helpers, and potentially include these at a later time.

The integration test is actually correctly checking for the shifting `rel` attribute, which is the symptom of the issue, meaning the test fails when [the actual fix](https://github.com/WordPress/gutenberg/pull/50460/files#diff-ea35a21c6148c935e1d6843190ef4ba37d443c1ca224a145a59b745304fb6664R40) is removed. Due to that, I think the integration test is worth retaining as it does help to identify the issue, and is not expensive to run. Let me know if you have any further thoughts about including (or excluding) the integration test before we merge. 🙇

dcalhoun reviewed (approved):
Thank you for adding the test coverage. 🙇🏻
